src/LVMGenerator.h

```diff
@@ -104,6 +176,146 @@ class RelationEncoder {
+    }
+
+    /**
+     * Return outer-most filter operations.
+     */
+    const std::vector<const RamCondition*>& getOuterFilterOperation() {
+        return outerOperations;
+    }
+
+    /**
+     * Return views for outer-most filter operations.
+     */
+    const std::vector<const RamNode*>& getViewsInOuterOperation() {
+        return viewsForFilterOperation;
+    }
+
+    /**
+     * Return nested operations
+     */
+    std::vector<const RamNode*> getViewsInNestedOperation() {
+        return viewsForNestedOperation;
+    }
+
+    /**
+     * Find all operations under the root node that require a view.
+     *
+     * In particular, find all index operation and existence check operation.
+     */
+    std::vector<const RamNode*> findAllViews(const RamNode& node) {
+        std::vector<const RamNode*> res;
+        visitDepthFirst(node, [&](const RamNode& node) {
+            if (requireView(&node) == true) {
+                res.push_back(&node);
+            };
+        });
+        return res;
+    }
+
+    /**
+     * Return true if the given operation requires a view.
+     */
+    bool requireView(const RamNode* node) {
+        if (auto exists = dynamic_cast<const RamExistenceCheck*>(node)) {
+            for (auto& v : exists->getValues()) {
+                if (!isRamUndefValue(v)) {
+                    return true;
+                }
+            }
+            return false;
+        } else if (auto provExists = dynamic_cast<const RamProvenanceExistenceCheck*>(node)) {
+            for (auto& v : provExists->getValues()) {
+                if (!isRamUndefValue(v)) {
+                    return true;
+                }
+            }
+            return false;
+        } else if (auto indexOp = dynamic_cast<const RamIndexOperation*>(node)) {
+            for (auto& v : indexOp->getRangePattern()) {
+                if (!isRamUndefValue(v)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+        return false;
+    }
+
+    void reset() {
+        outerOperations.clear();
+        viewsForFilterOperation.clear();
+        viewsForNestedOperation.clear();
+    }
+
+    /**
+     * Same as toConjunctionList defined in RamOperation. But does not clone new node,
+     * only holds a list of raw pointers to the original node.
+     */
+    inline std::vector<const RamCondition*> toConjunctionList(const RamCondition* condition) {
+        std::vector<const RamCondition*> list;
+        std::stack<const RamCondition*> stack;
+        if (condition != nullptr) {
+            stack.push(condition);
+            while (!stack.empty()) {
+                condition = stack.top();
+                stack.pop();
+                if (const auto* ramConj = dynamic_cast<const RamConjunction*>(condition)) {
+                    stack.push(&ramConj->getLHS());
+                    stack.push(&ramConj->getRHS());
+                } else {
+                    list.emplace_back(condition);
```

Comment:
Use push_back. emplace_back/push_back are equivalent here since we're dealing with pointers, but it makes it clearer that you're not making a new node.